### PR TITLE
R doc 1811 onbeforedelete

### PR DESCRIPTION
**Pull request by morhilai**

### CODE REVIEW

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+Entities can be marked for deletion by using the `Delete` method, but will not be removed from the server until `SaveChanges` is called.
+
+## Syntax
+
+{CODE deleting_1@ClientApi\Session\DeletingEntities.cs /}
+
+| Parameters | | |
+| ------------- | ------------- | ----- |
+| **entity** or **id** | T or string | instance of the entity to delete or entity ID |
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown:11]: separate rows

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+You can read more about defer operations [here](./how-to/defer-operations).
+
+{INFO/}
+
+## Related Articles
+
+### Session
+
+- [What is a Session and How Does it Work](../../client-api/session/what-is-a-session-and-how-does-it-work)
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown:58]: add link to UseOptimisticConcurrency 
`client-api/session/configuration/how-to-enable-optimistic-concurrency`

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+It should be defined with this signature:  
+
+{CODE-BLOCK: csharp}
+private void OnBeforeDeleteEvent(object sender, BeforeDeleteEventArgs args);
+{CODE-BLOCK/}
+
+| Parameters | Type | Description |
+| - | - | - |
+| **sender** | `IDocumentSession` | The session on which `SaveChanges()` has been called, triggering this event |
+| **args** | `BeforeDeleteEventArgs` | `args` contains the session on which `SaveChanges()` has been called, the ID of the document being deleted, and the document itself. |
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:41]:
```suggestion
| **args** | `BeforeDeleteEventArgs` | `args` contains the session on which `SaveChanges()` has been called, the ID of the document being deleted, and the document metadata and the document itself. |
```

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+The class `BeforeDeleteEventArgs`:
+
+{CODE-BLOCK: csharp}
+public class BeforeDeleteEventArgs
+{
+    public InMemoryDocumentSessionOperations Session;
+    public string DocumentId;
+    public object Entity;
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:50]: add 
 public IMetadataDictionary DocumentMetadata
internal bool MetadataAccessed

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+{PANEL/}
+
+{PANEL:OnAfterConversionToEntity}
+
+This event is invoked after conversion of a JSON document to an entity.
+
+{CODE on_after_conversion_to_entity@ClientApi\Session\Events.cs /}
+
+It takes the argument `AfterConversionToEntityEventArgs`, that consists of a JSON document, its ID, the session instance and a converted entity.
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:136]: move to line before the code example

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+After we subscribe to the event, every stored entity will invoke the method.
+
+{CODE store_session@ClientApi\Session\Events.cs /}
+
+{PANEL/}
+
+{PANEL:OnBeforeDelete}
+
+This event is invoked as a part of `SaveChanges`, but before it actually sends the deleted entities to the server.
```
Comment (garayx) [inline Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:31]: add that it works on `Delete(id)` and `Delete(entity)`

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown (hunk 1 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown:11]: Done

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown (hunk 2 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/deleting-entities.dotnet.markdown:58]: Done

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown (hunk 3 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:41]: Same for `Before*Store*EventArgs`?
If so, fixed for both

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown (hunk 4 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:50]: Same for `Before*Store*EventArgs`?
If so, fixed for both

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown (hunk 5 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:136]: Fixed

File: Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown (hunk 6 above)
Comment (morhilai) [reply Documentation/5.1/Raven.Documentation.Pages/client-api/session/how-to/subscribe-to-events.dotnet.markdown:31]: Done

**Review by garayx — APPROVED** [review]

**Pull request merged**